guide/src/class.md

```diff
@@ -716,6 +716,54 @@ impl MyClass {
 }
 ```
 
+## Free functions
+
+Free functions defined using `#[pyfunction]` interact with classes through the same mechanisms as the self parameters of instance methods, i.e. they can take GIL-bound references, GIL-bound reference wrappers or GIL-indepedent references:
+
+```rust
+# #![allow(dead_code)]
+# use pyo3::prelude::*;
+#[pyclass]
+struct MyClass {
+    my_field: i32,
+}
+
+#[pyfunction]
+fn increment_field(my_class: &mut MyClass) {
+    my_class.my_field += 1;
+}
+
+#[pyfunction]
+fn print_field(my_class: PyRef<'_, MyClass>) {
+    println!("{}", my_class.my_field);
+}
+
+#[pyfunction]
+fn print_refcnt(my_class: Py<MyClass>, py: Python<'_>) {
+    println!("{}", my_class.get_refcnt(py));
+}
+```
+
+Classes can also be passed by value if they can be be cloned, i.e. they automatically implement `FromPyObject` if they implement `Clone`, e.g. via `#[derive(Clone)]`:
+
+```rust
+# #![allow(dead_code)]
+# use pyo3::prelude::*;
+#[pyclass]
+#[derive(Clone)]
+struct MyClass {
+    my_field: Box<i32>,
+}
+
+#[pyfunction]
+fn dissamble_clone(my_class: MyClass) {
+    let MyClass { mut my_field } = my_class;
+    *my_field += 1;
+}
+```
+
+Note that `#[derive(FromPyObject)]` on a class is usually not useful as it tries to construct a new Rust value by filling in the fields by looking up attributes of any given Python value.
```

Comment:
Crazy thought - we might be able to guard against this footgun by emitting some code which fails to compile if `T: !Clone + FromPyObject`, which users could then turn off with a `#[pyclass(external_impl_frompyobject)]` setting.

(Would probably require some great specialization evils in the implementation, my favourite kind of madness 😂)

Reply:
I am not sure about the additional customization knob on `#[pyclass]`. In an ideal world, I would prefer to emit a hint when compilation fails due to a missing `FromPyObject` impl on a class, i.e. suggesting to derive `Clone` but not `FromPyObject`.

Reply:
Yes, I think further down the line this may be possible (e.g. maybe `#[diagnostic::on_unimplemented]`). For now let's not worry about this, unless we see lots of users suffer here.